src/chains/ethereum/ethereum/src/api.ts

```diff
@@ -2877,6 +2877,184 @@ export default class EthereumApi implements Api {
+        `0.${(
+          (currentBlock.header.gasUsed.toBigInt() * PRECISION_BIG_INT) /
+          currentBlock.header.gasLimit.toBigInt()
+        )
+          .toString()
+          .padStart(PAD_PRECISION, "0")}`
+      );
+
+      if (reward) {
+        const transactions = currentBlock.getTransactions();
+
+        // If there are no transactions, all reward percentiles are 0.
+        if (transactions.length === 0) {
+          reward[currentPosition] = rewardPercentiles.map(() => {
+            return Quantity.Zero;
+          });
+        } else {
+          // For all transactions, effectiveGasReward = normalized fee per unit of gas
+          // earned by the miner regardless of transaction type
+          const effectiveRewardAndGasUsed = (
+            await Promise.all(
+              transactions.map(async tx => {
+                const receipt = (
+                  await blockchain.transactionReceipts.get(tx.hash.toBuffer())
+                ).toJSON(currentBlock, tx, blockchain.common);
+
+                const baseFeePerGas = currentBlock.header.baseFeePerGas
+                  ? currentBlock.header.baseFeePerGas.toBigInt()
+                  : 0n;
+
+                let effectiveGasReward;
+                if ("maxPriorityFeePerGas" in tx) {
+                  effectiveGasReward =
+                    tx.maxFeePerGas.toBigInt() - baseFeePerGas;
+                  if (tx.maxPriorityFeePerGas < effectiveGasReward) {
+                    effectiveGasReward = tx.maxPriorityFeePerGas;
+                  }
+                } else {
+                  effectiveGasReward = tx.gasPrice.toBigInt() - baseFeePerGas;
+                }
+
+                return {
+                  effectiveGasReward: Quantity.toBigInt(effectiveGasReward),
+                  gasUsed: receipt.gasUsed ? receipt.gasUsed.toBigInt() : 0n
+                };
+              })
+            )
+          ).sort((a, b) => {
+            if (a.effectiveGasReward > b.effectiveGasReward) return 1;
+            if (a.effectiveGasReward < b.effectiveGasReward) return -1;
+            return 0;
+          });
+
+          // All of the block transactions are ordered, ascending, from least to greatest by
+          // the fee the tx paid per unit of gas. For each percentile of block gas consumed,
+          // what was the fee paid for the unit of gas at that percentile.
+          reward[currentPosition] = rewardPercentiles.map(percentile => {
+            let gasUsed = 0n;
+
+            const targetGas =
+              (currentBlock.header.gasUsed.toBigInt() *
+                BigInt(percentile * PRECISION_FLOAT)) /
+              PRECISION_BIG_INT;
+
+            for (const values of effectiveRewardAndGasUsed) {
+              gasUsed = gasUsed + values.gasUsed;
+
+              if (targetGas <= gasUsed) {
+                return Quantity.from(values.effectiveGasReward);
+              }
+            }
+
+            return Quantity.from(
```

Comment:
Will this ever happen? Shouldn't the gas off all of the transactions always sum up to the block gas used?

I don't think this line can happen, but it _did_ make me realize an optimization - if the `percentile` is `100`, won't that always mean it's the last transaction in this list? You could check chat and return `effectiveRewardAndGasUsed[effectiveRewardAndGasUsed.length - 1]` up front and not have to loop through all of them.

Reply:
Oh, but also, `targetGas` can be > 100% of gas used if a percentile over 100 is given.

Reply:
Has this one been resolved?

Reply:
Yes, recent changes reject `rewardPercentiles` that are not monotonic and for percentiles above 100. This matches the Infura API.

Because percentiles can never go above 100 this line (correctly as @MicaiahReid points out) becomes dead code and has been removed.